pyasn1/codec/ber/decoder.py

```diff
@@ -303,24 +156,45 @@ def valueDecoder(self, substrate, asn1Spec,
                      **options):
 
         if substrateFun:
-            return substrateFun(self._createComponent(
-                asn1Spec, tagSet, noValue, **options), substrate, length)
+            asn1Object = self._createComponent(asn1Spec, tagSet, noValue, **options)
 
-        if not length or _endOfStream(substrate):
+            for chunk in substrateFun(asn1Object, substrate, length):
+                yield chunk
+
+            return
+
+        if not length:
+            raise error.PyAsn1Error('Empty BIT STRING substrate')
+
+        for chunk in streaming.isEndOfStream(substrate):
```

Comment:
Performance nit-pick: I'd import all names explicitly (to be applied in here and in other places) so that you don't have to access the module's namespace.

On my computer with Python 3.7, the extra dot adds 20 ns per call.

Reply:
https://github.com/janpipek/various_benchmarks/blob/master/python/access_penalties.ipynb